### PR TITLE
fix: convert age and idle fields to timestamps

### PR DESCRIPTION
**Pull request by h3isenbug**
Fix for https://github.com/oliver006/redis_exporter/issues/496

### CODE REVIEW

**Comment by h3isenbug** [comment]
@oliver006

File: exporter/clients.go [hunk 1]
```diff
@@ -1,12 +1,13 @@
 package exporter
 
 import (
-	"regexp"
-	"strings"
-
 	"github.com/gomodule/redigo/redis"
 	"github.com/prometheus/client_golang/prometheus"
 	log "github.com/sirupsen/logrus"
+	"regexp"
```
Comment (oliver006) [inline exporter/clients.go:7]: Can you fix the sorting again? stdlibs first, then a blank line, then 3rd party imports.

File: exporter/clients.go [hunk 2]
```diff
@@ -28,15 +29,27 @@ func parseClientListString(clientInfo string) ([]string, bool) {
...
 	hostPortString := strings.Split(connectedClient["addr"], ":")
 	if len(hostPortString) != 2 {
 		return nil, false
 	}
 
 	return []string{
 		connectedClient["name"],
-		connectedClient["age"],
-		connectedClient["idle"],
+		strconv.FormatInt(createdAt, 10),
```
Comment (oliver006) [inline exporter/clients.go:51]: Can you use `strconv.Itoa()` ? And call that in the `durationFieldToTimestamp()` helper function?

File: exporter/clients.go [hunk 3]
```diff
@@ -48,6 +61,16 @@ func parseClientListString(clientInfo string) ([]string, bool) {
 
 }
 
+func durationFieldToTimestamp(field string) (int64, error) {
+	parsed, err := strconv.ParseInt(field, 10, 64)
```
Comment (oliver006) [inline exporter/clients.go:65]: strconv.Atoi() is the more commonly used function, can you use that?

File: exporter/clients_test.go [hunk 4]
```diff
@@ -1,16 +1,24 @@
...
 import (
+	"strconv"
 	"strings"
 	"testing"
 
 	"github.com/prometheus/client_golang/prometheus"
 )
 
 func TestParseClientListString(t *testing.T) {
+	createdAt6321, _ := durationFieldToTimestamp("6321")
```
Comment (oliver006) [inline exporter/clients_test.go:12]: Instead of having these called one-off here can you add a new `Test...` function that loops through a bunch of test cases and directly calls and tests `durationFieldToTimestamp()` ?

File: exporter/exporter.go [hunk 5]
```diff
@@ -306,7 +306,7 @@ func NewRedisExporter(redisURI string, opts Options) (*Exporter, error) {
 
 	e.metricDescriptions = map[string]*prometheus.Desc{}
 
-	connectedClientsLabels := []string{"name", "age", "idle", "flags", "db", "omem", "cmd", "host"}
+	connectedClientsLabels := []string{"name", "created_at", "idle_since", "flags", "db", "omem", "cmd", "host"}
```
Comment (oliver006) [inline exporter/exporter.go:309]: 👍

File: exporter/clients_test.go (hunk 4 above)
Comment (oliver006) [reply exporter/clients_test.go:12]: (That would/should also address the comments left by the bot about the lines not covered by tests)

**Comment by oliver006** [comment]
@h3isenbug  - just a friendly nudge, would love to see this merged as this could significantly reduce load on a prometheus server.

**Comment by h3isenbug** [comment]
@oliver006 Sorry for the delay. I'm very busy at the moment(getting married 😅).
I'll fix these as soon as I get a chance.

**Comment by oliver006** [comment]
Congratulations on getting married 🥳 !
Take your time and resume this whenever you're ready - there's no rush.

**Comment by oliver006** [comment]
@h3isenbug - are you still working on this?

**Comment by h3isenbug** [comment]
@oliver006 I will fix these and rebase it on master today 😊

File: exporter/clients.go (hunk 2 above)
Comment (h3isenbug) [reply exporter/clients.go:51]: Itoa accepts int, not int64 which is required to prevent int32 overflow in the years to follow.

Comment (h3isenbug) [reply exporter/clients.go:51]: purpose of `durationFieldToTimestamp` is to convert duration to timestamp, not to string. It makes sense to do FormatInt in the parseClientListString, because the logic of createdAt and idleSince being string is related to `parseClientListString`, not `durationFieldToTimestamp`

File: exporter/clients.go (hunk 3 above)
Comment (h3isenbug) [reply exporter/clients.go:65]: as described before, FormatInt and ParseInt are better fits for this purpose.

File: exporter/clients_test.go (hunk 4 above)
Comment (h3isenbug) [reply exporter/clients_test.go]: proper testing of `durationFieldToTimestamp` would require frozen time. I think this function is small enough to ignore it being not tested.
I improved `TestParseClientListString` as much as possible.

**Comment by h3isenbug** [comment]
@oliver006 I answered all of the discussions on this pr

File: exporter/clients.go (hunk 2 above)
Comment (oliver006) [reply exporter/clients.go:52]: Ah right, int64 compatibility is a good call. Still, can you do the call to `strconv.FormatInt()` inside durationFieldToTimestamp() and return a string instead of int? In that way you can also get rid of the helper function in the tests.

File: exporter/clients_test.go [hunk 6]
```diff
@@ -1,16 +1,28 @@
...
 import (
+	"strconv"
 	"strings"
 	"testing"
 
 	"github.com/prometheus/client_golang/prometheus"
 )
 
 func TestParseClientListString(t *testing.T) {
+	convertDurationToTimestampString := func(duration string) string {
```
Comment (oliver006) [inline exporter/clients_test.go:12]: See above, If you make `durationFieldToTimestamp()` return a string then we won't need this function.

**Review by oliver006 — COMMENTED** [review]
Good point about int64 - that makes sense.
I think you can convert the int64 result to a string and return that in `durationFieldToTimestamp()` - otherwise this is good to go!

File: exporter/clients_test.go (hunk 6 above)
Comment (h3isenbug) [reply exporter/clients_test.go:12]: returning string is not purpose of `durationFieldToTimestamp`. This would violate SRP.

Comment (oliver006) [reply exporter/clients_test.go:12]: Not sure what SRP is.
But you always convert the int that's returned from `durationFieldToTimestamp()` to a string so instead of doing that in reach place I prefer you'd just have the function return a string and then the caller can use it right away.

File: exporter/clients_test.go (hunk 4 above)
Comment (oliver006) [reply exporter/clients_test.go]: What I mean is using table driven tests which are common in Go.
Add tests for all the corner cases to improve test coverage.

See e.g. this one here in the std lib: https://golang.org/src/time/time_test.go#L1003

and this longer explanation: https://mj-go.in/golang/table-driven-tests-in-go

File: exporter/clients_test.go (hunk 6 above)
Comment (zuiderkwast) [reply exporter/clients_test.go:12]: SRP = single-responsibility principle

(That's what I assume it means here.)

Maybe just rename durationFieldToTimestamp to durationFieldToTimestampString and make it responsible for the whole conversion? (It's still a single responsibility.)

Comment (test12345123456) [reply exporter/clients_test.go:12]: @zuiderkwast That would not make it comply with SRP, but I don't have enough time to convince everyone on this matter. I'll do it as soon as I get the time to do so. If anyone had some spare time to fix this before I do, go ahead.

Comment (h3isenbug) [reply exporter/clients_test.go:12]: @zuiderkwast That would not make it comply with SRP, but I don't have enough time to convince everyone on this matter. I'll do it as soon as I get the time to do so. If anyone had some spare time to fix this before I do, go ahead.

**Comment by oliver006** [comment]
@h3isenbug - gentle nudge, are you still working on this? If not then please let me know and I'll try to salvage what I can from this PR. Thanks!

**Comment by h3isenbug** [comment]
@oliver006 I've been very busy lately and I think it would be great if you continued stuff done on this pr. Thanks.

**Comment by oliver006** [comment]
Finishing this in #534 - thanks for getting this started @h3isenbug !
